csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -199,31 +427,98 @@ std::pair<int64_t, int64_t> getPersistentBufferSize(
-  // reductions must be persistent, allow register spill avoid segmentation
-  std::vector<TensorView*> outer_reduction_tvs;
-  for (auto tv : reduction_tvs) {
-    if (!scheduler_utils::isFastestDimReduction(tv)) {
-      outer_reduction_tvs.emplace_back(tv);
+  const auto& persistent_buffers = buffer_params.project_to_input
+      ? persistent_buffer_info.projectable_buffer_inputs
+      : persistent_buffer_info.persistent_buffers;
+
+  auto total_buffer_size = buffer_params.project_to_input
+      ? persistent_buffer_size_info.projected_persistent_buffer_size
+      : persistent_buffer_size_info.persistent_buffer_size;
+  total_buffer_size +=
+      partialOuterReductionBufferSize(reduction_tvs, runtime_info);
+
+  const auto dev_prop = at::cuda::getCurrentDeviceProperties();
+  int64_t smem_overhead =
+      scheduler_utils::getSharedMemoryOverheadPerBlock(fusion, reduction_tvs);
+  int64_t available_smem =
+      (int64_t)dev_prop->sharedMemPerMultiprocessor - smem_overhead;
+  int64_t available_regs = scheduler_utils::register_file_size_full;
+  buffer_params.smem_overhead = smem_overhead;
+
+  // Put all the persistent tensors in registers
+  buffer_params.regs_buffer_size = total_buffer_size;
+  buffer_params.smem_buffer_size = 0;
+
+  // Relocate buffers to shared memory until the buffer size in registers is
+  // within the allowable limit.
+  if (buffer_params.regs_buffer_size > available_regs) {
+    const auto sorted_candidate_tvs = sortPersistentBuffers(
+        fusion,
+        persistent_buffers,
+        persistent_buffer_info.projectable_buffer_inputs,
+        reduction_tvs,
+        runtime_info);
+    // calculate the accumulated buffer size of the first N buffers
+    const int64_t n_buffers = (int64_t)persistent_buffers.size();
+    std::vector<int64_t> acc_regs_buffer_sizes(n_buffers + 1, 0);
+    std::vector<int64_t> acc_smem_buffer_sizes(n_buffers + 1, 0);
+    for (int i = 1; i <= n_buffers; i++) {
+      int64_t tv_buffer_size_regs =
+          scheduler_utils::getPersistentBufferSizeOfTensor(
+              sorted_candidate_tvs[i - 1],
+              runtime_info,
+              persistent_buffer_info);
+      int64_t tv_buffer_size_smem = roundUpSharedMemory(
+          sorted_candidate_tvs[i - 1],
+          tv_buffer_size_regs,
+          vectorize_factor,
+          InnerOuterPersistentKernelScheduler::threads_per_block_min,
+          InnerOuterPersistentKernelScheduler::threads_per_block_max,
+          dev_prop->warpSize);
+
+      acc_regs_buffer_sizes[i] =
+          acc_regs_buffer_sizes[i - 1] + tv_buffer_size_regs;
+      acc_smem_buffer_sizes[i] =
+          acc_smem_buffer_sizes[i - 1] + tv_buffer_size_smem;
+    }
+
+    // Determine the least number of buffers to transfer to shared memory
+    // to ensure the register buffer size doesn't exceed the available limit.
+    int64_t n_smem_buffer = -1;
+    for (int i = 1; i <= n_buffers; i++) {
```

Comment:
nitpick: this for loop could go inside `if (buffer_params.regs_buffer_size > available_regs) {`

Reply:
^^^ wait.. I probably read this one wrong. The logic here means that we'll *always* give at least 1 buffer to smem. Is that intended? if `buffer_params.regs_buffer_size <= available_regs`, shouldn't all buffer go to register instead?

Reply:
I think you may read it wrong. This part is nested inside `if (buffer_params.regs_buffer_size > available_regs) {`, so if condition is false, all buffers go to register.